lib/wallet/coinselector.js

```diff
@@ -0,0 +1,897 @@
+      if (!this.isFull())
+        this.fund();
+    } while (!this.isFull() && this.index < this.coins.length);
+  }
+
+  /**
+   * Initiate selection based on a hard fee.
+   */
+
+  selectHard() {
+    this.fee = Math.min(this.hardFee, CoinSelector.MAX_FEE);
+    this.fund();
+  }
+
+  /**
+   * Initiate multi algorithm coin selection.
+   */
+
+  async attemptSelection() {
+    // calculate cost of creating and spending a change output
+    // add dummy output for change.
+    const change = new Output();
+
+    if (this.changeAddress) {
+      change.script.fromAddress(this.changeAddress);
+    } else {
+      // In case we don't have a change address,
+      // we use a fake p2wpkh output to gauge size.
+      change.script.fromProgram(0, Buffer.allocUnsafe(20));
+    }
+
+    const changeOutputSize = change.getSize();
+    const changeOutputCoin = new Coin();
+    changeOutputCoin.script = change.script;
+    const changeSpendingSize =
+      await changeOutputCoin.estimateSpendingSize(this.getAccount);
+    const costOfSpendingChange =
+      changeSpendingSize * CoinSelector.LONG_TERM_FEERATE;
+
+    // cost of change = fee paid to create change +
+    // fee paid to spend that change in future
+    const costOfChange = this.getFee(changeOutputSize) + costOfSpendingChange;
+
+    let target = this.outputValue;
+    let size = await this.tx.estimateSize(this.getAccount);
+    const fee = this.getFee(size);
+    target += fee;
+
+    let bestSelection = [];
+    let bestWaste = consensus.MAX_MONEY;
+    let isBnB = false; // true if bestSelection is BnB
+
+    // try to find a changeless solution using BnB
+    let result = this.selectBnB(target, costOfChange);
+    if (result && result.length > 0) {
+      bestSelection = result;
+      bestWaste = this.getWaste(result, costOfChange, target);
+      isBnB = true;
+    }
+
+    // fallback to other algorithms as we can't find a changeless solution
+    // update target to include cost of producing change
+    target += this.getFee(changeOutputSize);
+
+    // find a solution of target using Lowest Larger
+    result = this.selectLowestLarger(target);
+    if (result.length > 0) {
+      const currWaste = this.getWaste(result, costOfChange, target);
+      if (currWaste < bestWaste) {
+        bestSelection = result;
+        bestWaste = currWaste;
+      }
+    }
+
+    // find a solution of target using SRD
+    result = this.selectSRD(target);
+    if (result.length > 0) {
+      const currWaste = this.getWaste(result, costOfChange, target);
+      if (currWaste < bestWaste) {
+        bestSelection = result;
+        bestWaste = currWaste;
+      }
+    }
+
+    if (bestSelection && bestSelection.length > 0) {
+      for (const i of bestSelection) {
+        const coin = this.coins[this.coinPointers[i].index];
+        this.tx.addCoin(coin);
+        this.chosen.push(coin);
+      }
+      // add change output if solution is not found by BnB
+      if (!isBnB)
```

Comment:
Other coin selection algorithms could also produce changeless solutions (although that's less likely), so perhaps:

```suggestion
      if (!isChangeless)
```

Reply:
yeah, I am really bad at naming things. I'll improve.